File: ios/WhiteNoise/WhiteNoise/PlaybackController.swift
import Combine
import SwiftUI
import UIKit

/// Coordinates white-noise playback with battery-aware volume reduction.
@MainActor
final class PlaybackController: ObservableObject {
    @Published var isPlaying = false
    @Published var userVolume: Double = 0.55
    @Published var effectiveVolume: Double = 0.55
    @Published var batteryMultiplier: Double = 1.0
    @Published var statusMessage: String = "Ready"
    @Published var errorMessage: String?

    let battery = BatteryMonitor()
    private let engine = WhiteNoiseEngine()
    private var cancellables = Set<AnyCancellable>()

    /// Battery thresholds → volume multipliers (applied only while unplugged).
    /// Above 30%: full user volume.
    /// 20–30%: gentle trim.
    /// 10–20%: stronger trim.
    /// Below 10%: deep trim to stretch remaining charge.
    private let thresholds: [(minLevel: Float, multiplier: Double)] = [
        (0.30, 1.00),
        (0.20, 0.80),
        (0.10, 0.55),
        (0.00, 0.30)
    ]

    init() {
        battery.$level
            .combineLatest(battery.$state, $userVolume)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _, _, _ in
                self?.applyVolumePolicy()
            }
            .store(in: &cancellables)

        applyVolumePolicy()
    }

    func togglePlayback() {
        if isPlaying {
            stop()
        } else {
            start()
        }
    }

    func start() {
        errorMessage = nil
        do {
            try engine.configureSession()
            applyVolumePolicy()
            try engine.start()
            isPlaying = true
            UIApplication.shared.isIdleTimerDisabled = true
            statusMessage = "Playing indefinitely"
        } catch {
            isPlaying = false
            errorMessage = error.localizedDescription
            statusMessage = "Could not start audio"
        }
    }

    func stop() {
        engine.stop()
        isPlaying = false
        UIApplication.shared.isIdleTimerDisabled = false
        statusMessage = "Paused"
    }

    private func applyVolumePolicy() {
        let multiplier = computeBatteryMultiplier()
        batteryMultiplier = multiplier
        let effective = userVolume * multiplier
        effectiveVolume = effective
        // Map UI 0...1 into a comfortable noise gain range.
        let gain = Float(effective) * 0.55
        engine.updateGain(gain)

        if battery.isCharging {
            statusMessage = isPlaying
                ? "Playing · charging — full volume"
                : "Ready · charging"
        } else if multiplier < 1 {
            let pct = Int((battery.level * 100).rounded())
            statusMessage = isPlaying
                ? "Playing · battery \(pct)% — volume reduced"
                : "Ready · battery \(pct)% — volume will reduce"
        } else if isPlaying {
            statusMessage = "Playing indefinitely"
        }
    }

    private func computeBatteryMultiplier() -> Double {
        // While charging or full, never throttle — the phone is plugged in.
        if battery.isCharging {
            return 1.0
        }
        let level = battery.level
        for threshold in thresholds {
            if level >= threshold.minLevel {
                return threshold.multiplier
            }
        }
        return 0.30
    }
}

File: hush/js/core/playbackController.js
import {
  batteryDetail,
  batteryPercentLabel,
  batteryVolumeMultiplier,
  effectiveGain,
  effectiveVolume
} from './batteryPolicy.js';

/**
 * Coordinates playback with battery-aware volume.
 * Depends only on AudioPort + BatteryPort — no DOM.
 */
export class PlaybackController {
  /**
   * @param {import('../ports/audioPort.js').AudioPort} audio
   * @param {import('../ports/batteryPort.js').BatteryPort} battery
   * @param {{ userVolume?: number, onChange?: (state: ReturnType<PlaybackController['getState']>) => void }} [options]
   */
  constructor(audio, battery, options = {}) {
    this._audio = audio;
    this._battery = battery;
    this._userVolume = clampVolume(options.userVolume ?? 0.55);
    this._onChange = typeof options.onChange === 'function' ? options.onChange : null;
    this._isPlaying = false;
    this._errorMessage = null;
    this._unsub = null;

    this._unsub = this._battery.subscribe(() => this._applyVolumePolicy());
    this._applyVolumePolicy();
  }

  dispose() {
    if (this._unsub) {
      this._unsub();
      this._unsub = null;
    }
  }

  getState() {
    const battery = this._battery.getSnapshot();
    const multiplier = batteryVolumeMultiplier(battery);
    return {
      isPlaying: this._isPlaying,
      userVolume: this._userVolume,
      batteryMultiplier: multiplier,
      effectiveVolume: effectiveVolume(this._userVolume, multiplier),
      battery,
      batteryPercent: batteryPercentLabel(battery),
      batteryDetail: batteryDetail(battery),
      statusMessage: this._statusMessage(battery, multiplier),
      errorMessage: this._errorMessage
    };
  }

  async toggle() {
    if (this._isPlaying) {
      this.stop();
    } else {
      await this.start();
    }
  }

  async start() {
    this._errorMessage = null;
    try {
      this._applyVolumePolicy();
      await this._audio.start();
      this._isPlaying = true;
      this._emit();
    } catch (err) {
      this._isPlaying = false;
      this._errorMessage = err && err.message ? err.message : String(err);
      this._emit();
    }
  }

  stop() {
    this._audio.stop();
    this._isPlaying = false;
    this._emit();
  }

  /**
   * @param {number} value 0..1
   */
  setUserVolume(value) {
    this._userVolume = clampVolume(value);
    this._applyVolumePolicy();
  }

  _applyVolumePolicy() {
    const battery = this._battery.getSnapshot();
    const multiplier = batteryVolumeMultiplier(battery);
    const gain = effectiveGain(this._userVolume, multiplier);
    this._audio.setGain(gain);
    this._emit();
  }

  _statusMessage(battery, multiplier) {
    if (this._errorMessage) return 'Could not start audio';
    if (battery.charging) {
      return this._isPlaying ? 'Playing · charging — full volume' : 'Ready · charging';
    }
    if (multiplier < 1) {
      const pct = batteryPercentLabel(battery);
      return this._isPlaying
        ? `Playing · battery ${pct} — volume reduced`
        : `Ready · battery ${pct} — volume will reduce`;
    }
    return this._isPlaying ? 'Playing indefinitely' : 'Ready';
  }

  _emit() {
    if (this._onChange) this._onChange(this.getState());
  }
}

function clampVolume(n) {
  if (typeof n !== 'number' || Number.isNaN(n)) return 0.55;
  return Math.min(1, Math.max(0.05, n));
}
